### PR TITLE
Update workload controller for new design

### PR DESCRIPTION
**Pull request by hossainemruz**
**Requirements:**
- [x] Merge https://github.com/appscode/stash/pull/647 to master
- [x] Merge https://github.com/appscode/stash/pull/673 to master
- [x] Merge https://github.com/appscode/stash/pull/674 to master
- [x] Merge https://github.com/appscode/stash/pull/671 to master
- [x] Rebase against master

**Tasks:**
- [x] Update Restore Logic
- [x] Update Backup Logic
- [x] Update Workload Controllers
	- [x] Deployment
	- [x] DaemonSet
	- [x] ReplicationControler
	- [x] ReplicaSet
	- [x] StatefulSet
- [x] Fix ConfigMap lock deletion and RBAC stuff deletion logic
- [x] Ensure sidecar/init-container according to new design
	- [x] Ensure Restore init-container injection
	- [x] Ensure Restore init-container deletion
	- [x] Ensure Backup sidecar injection
	- [x] Ensure Backup sidecar deletion
- [x] Add new backup command (sagor is working)
- [x] Add new restore command

- [x] Add controller (possibly in different PR)
	- [x] PersistentVolumeClaim
	- [x] AppBinding

### CODE REVIEW

File: apis/stash/v1alpha1/annotations.go [hunk 1]
```diff
@@ -1,9 +1,8 @@
...
 
 const (
-	ResticKey                = "restic.appscode.com"
-	LastAppliedConfiguration = ResticKey + "/last-applied-configuration"
-	VersionTag               = ResticKey + "/tag"
+	ResticKey         = "restic.appscode.com"
+	LastAppliedRestic = ResticKey + "/last-applied-configuration"
 	// ResourceVersion will be used to trigger restarts for ReplicaSet and RC pods
-	ResourceHash = ResticKey + "/resource-hash"
+	AppliedResticSpecHash = ResticKey + "/spec-hash"
```
Comment (tamalsaha) [inline apis/stash/v1alpha1/annotations.go]: We can't break the v1alpha1 release. If you need, add a new annotation key.

File: client/clientset/versioned/typed/stash/v1beta1/util/backupsession.go [hunk 2]
```diff
@@ -18,6 +18,9 @@ import (
 
 func CreateOrPatchBackupSession(c cs.StashV1beta1Interface, meta metav1.ObjectMeta, transform func(bs *api.BackupSession) *api.BackupSession) (*api.BackupSession, kutil.VerbType, error) {
 	cur, err := c.BackupSessions(meta.Namespace).Get(meta.Name, metav1.GetOptions{})
+	if err != nil {
```
Comment (tamalsaha) [inline client/clientset/versioned/typed/stash/v1beta1/util/backupsession.go]: Remove this println

File: apis/stash/v1beta1/annotations.go [hunk 3]
```diff
@@ -1,10 +1,23 @@
...
 const (
-	BackupKey                = "backup.appscode.com"
-	LastAppliedConfiguration = BackupKey + "/last-applied-configuration"
-	VersionTag               = BackupKey + "/tag"
+	VersionTag = BackupKey + "/tag"
 	// ResourceVersion will be used to trigger restarts for ReplicaSet and RC pods
-	ResourceHash = BackupKey + "/resource-hash"
-	StashKey     = "stash.appscode.com"
+	StashKey                       = "stash.appscode.com"
+	BackupKey                      = "backup.appscode.com"
```
Comment (tamalsaha) [inline apis/stash/v1beta1/annotations.go]: All keys should use the one single prefix `stash.appscode.com`

File: docs/examples/databases/postgres/backup_template.yaml [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+metadata:
+  name: pg-backup
+spec:
+  backend:
+    local:
+      mountPath: /safe/data
+      hostPath:
+        path: /data/stash-test/${TARGET_NAMESPACE}/${TARGET_KIND}/${TARGET_NAME}
+    storageSecretName: local-secret
+  schedule: "* * * * *"
```
Comment (tamalsaha) [inline docs/examples/databases/postgres/backup_template.yaml]: Either single or double quote. Better stay consistent.

File: docs/examples/databases/postgres/default_backup.yaml [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    kubedb.com/name: quick-postgres
+  annotations:
+    stash.appscode.com/backup-template: pg-backup
+spec:
+  clientConfig:
+    insecureSkipTLSVerify: true
+    service:
+      name: quick-postgres
+      port: 5432
+      scheme: "http"
```
Comment (tamalsaha) [inline docs/examples/databases/postgres/default_backup.yaml]: I usually avoid unnecessary quotes.

File: pkg/util/backup_configuration.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+	// keep only those BackupConfiguration that has this workload as target
+	for _, bc := range backupConfigurations {
+		if bc.DeletionTimestamp == nil && IsTarget(bc.Spec.Target, w) {
+			result = append(result, bc)
+		}
+	}
+
+	// if there is more than one BackupConfiguration then return error
+	if len(result) > 1 {
+		var msg bytes.Buffer
```
Comment (diptadas) [inline pkg/util/backup_configuration.go]: use string

File: pkg/controller/backup_session.go [hunk 7]
```diff
@@ -144,41 +183,77 @@ func (c *StashController) executeBackupSession(backupSession *api.BackupSession)
...
+		in.Spec.Template.Spec.ServiceAccountName = serviceAccountName
+
+		return in
 	})
-	return job, err
+
+	return err
+}
+
+func (c *StashController) setBackupSessionFailed(backupSession *api_v1beta1.BackupSession, jobErr error) error {
```
Comment (diptadas) [inline pkg/controller/backup_session.go]: use `UpdatePostBackupStatus` from status package

File: apis/stash/v1beta1/annotations.go (hunk 3 above)
Comment (hossainemruz) [reply apis/stash/v1beta1/annotations.go]: fixed in https://github.com/appscode/stash/pull/701

**Pull request merged**